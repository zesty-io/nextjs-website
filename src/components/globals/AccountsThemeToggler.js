import React from 'react';
import Button from '@mui/material/Button';
import { useTheme } from '@mui/material/styles';

import DarkModeIcon from '@mui/icons-material/DarkMode';
import LightModeIcon from '@mui/icons-material/LightMode';
import { grey } from '@mui/material/colors';

export const AccountsThemeToggler = () => {
  const theme = useTheme();
  const { themeToggler } = theme;
  const { mode } = theme.palette;

  return (
    <Button
      variant={'outlined'}
      onClick={() => themeToggler()}
      aria-label="Dark mode toggler"
      color={mode === 'light' ? 'inherit' : 'primary'}
      title={mode === 'light' ? 'Dark Mode' : 'Light Mode'}
      sx={{
        borderRadius: 1,
        minWidth: 'auto',
        padding: 0.7,
        borderColor: grey[200],
      }}
    >
      {mode === 'light' ? (
        <DarkModeIcon fontSize="small" color="disabled" />
      ) : (
        <LightModeIcon fontSize="small" />
      )}
    </Button>
  );
};
