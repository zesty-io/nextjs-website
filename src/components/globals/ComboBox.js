import * as React from 'react';
import PropTypes from 'prop-types';
import TextField from '@mui/material/TextField';
import Autocomplete, {
  autocompleteClasses,
  createFilterOptions,
} from '@mui/material/Autocomplete';
import useMediaQuery from '@mui/material/useMediaQuery';
import ListSubheader from '@mui/material/ListSubheader';
import Popper from '@mui/material/Popper';
import { useTheme, styled } from '@mui/material/styles';
import { VariableSizeList } from 'react-window';
import Typography from '@mui/material/Typography';

const LISTBOX_PADDING = 8; // px

function renderRow(props) {
  const { data, index, style } = props;
  const dataSet = data[index];
  const inlineStyle = {
    ...style,
    top: style.top + LISTBOX_PADDING,
  };

  if (dataSet.hasOwnProperty('group')) {
    return (
      <ListSubheader key={dataSet.key} component="div" style={inlineStyle}>
        {dataSet.group}
      </ListSubheader>
    );
  }

  return (
    <Typography component="li" {...dataSet[0]} noWrap style={inlineStyle}>
      {dataSet[1]}
    </Typography>
  );
}

const OuterElementContext = React.createContext({});

const OuterElementType = React.forwardRef((props, ref) => {
  const outerProps = React.useContext(OuterElementContext);
  return <div ref={ref} {...props} {...outerProps} />;
});

function useResetCache(data) {
  const ref = React.useRef(null);
  React.useEffect(() => {
    if (ref.current != null) {
      ref.current.resetAfterIndex(0, true);
    }
  }, [data]);
  return ref;
}

// Adapter for react-window
const ListboxComponent = React.forwardRef(
  function ListboxComponent(props, ref) {
    const { children, ...other } = props;
    const itemData = [];
    children.forEach((item) => {
      itemData.push(item);
      itemData.push(...(item.children || []));
    });

    const theme = useTheme();
    const smUp = useMediaQuery(theme.breakpoints.up('sm'), {
      noSsr: true,
    });

    const itemCount = itemData.length;
    const itemSize = smUp ? 36 : 48;

    const getChildSize = (child) => {
      if (child.hasOwnProperty('group')) {
        return 48;
      }

      return itemSize;
    };

    const getHeight = () => {
      if (itemCount > 8) {
        return 8 * itemSize;
      }
      return itemData.map(getChildSize).reduce((a, b) => a + b, 0);
    };

    const gridRef = useResetCache(itemCount);

    return (
      <div ref={ref}>
        <OuterElementContext.Provider value={other}>
          <VariableSizeList
            itemData={itemData}
            height={getHeight() + 2 * LISTBOX_PADDING}
            width="100%"
            ref={gridRef}
            outerElementType={OuterElementType}
            innerElementType="ul"
            itemSize={(index) => getChildSize(itemData[index])}
            overscanCount={5}
            itemCount={itemCount}
          >
            {renderRow}
          </VariableSizeList>
        </OuterElementContext.Provider>
      </div>
    );
  },
);

ListboxComponent.propTypes = {
  children: PropTypes.node,
};

// function random(length) {
//   const characters =
//     'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
//   let result = '';

//   for (let i = 0; i < length; i += 1) {
//     result += characters.charAt(Math.floor(Math.random() * characters.length));
//   }

//   return result;
// }

const StyledPopper = styled(Popper)({
  [`& .${autocompleteClasses.listbox}`]: {
    boxSizing: 'border-box',
    '& ul': {
      padding: 0,
      margin: 0,
    },
  },
});

const filterOptions = createFilterOptions({
  matchFrom: 'any',
  stringify: (option) => option.name + option.value,
});

const Index = ({
  instances = [],
  setCookies,
  instanceZUID,
  width = 300,
  initialLabel,
  ...props
}) => {
  const theme = useTheme();
  const [label, setlabel] = React.useState('');
  const memoizedInstances = React.useMemo(() => {
    return instances;
  }, [instances]);

  const OPTIONS = memoizedInstances?.map((e) => {
    return {
      name: e.name,
      value: e.ZUID,
    };
  });

  const currentVal = memoizedInstances?.find((e) => e.ZUID === instanceZUID);

  const handleChange = (e, newValue) => {
    setCookies(newValue?.value);
    setlabel(newValue?.name);
  };

  React.useEffect(() => {
    instances && setlabel(currentVal?.name);
  }, [instances]);

  return (
    <Autocomplete
      filterOptions={filterOptions}
      id="virtualize-demo"
      onChange={handleChange}
      sx={{
        width,
        backgroundColor:
          theme.palette.mode === 'light' && theme.palette.common.white,
      }}
      disableListWrap
      PopperComponent={StyledPopper}
      ListboxComponent={ListboxComponent}
      options={OPTIONS}
      getOptionLabel={(option) => option.name}
      renderInput={(params) => {
        return (
          <TextField
            {...params}
            label={label || initialLabel || 'Select an instance'}
          />
        );
      }}
      renderOption={(props, option) => {
        return [props, option.name];
      }}
      {...props}
    />
  );
};
export const ComboBox = React.memo(Index);
