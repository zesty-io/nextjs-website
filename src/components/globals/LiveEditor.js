import { ZestyLiveEditor } from '@zesty-io/live-editor';

import React, { useEffect } from 'react';

const LiveEditor = ({ data }) => {
  useEffect(() => {
    ZestyLiveEditor(data);
  }, []);

  return <div></div>;
};

export default LiveEditor;
