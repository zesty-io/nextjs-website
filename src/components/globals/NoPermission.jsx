import { useMemo } from 'react';
import {
  Stack,
  Box,
  Typography,
  Avatar,
  List,
  ListItem,
  ListItemText,
  ListItemAvatar,
} from '@mui/material';

import { hashMD5 } from 'utils/Md5Hash';
import shield from '../../../public/assets/images/shield.svg';

export const NoPermission = ({ users }) => {
  const ownersAndAdmins = useMemo(() => {
    if (!users && !users?.length) return [];

    const owners = users
      .filter((user) => user.role?.name?.toLowerCase() === 'owner')
      .sort((a, b) => a.firstName.localeCompare(b.firstName));
    const admins = users
      .filter((user) => user.role?.name?.toLowerCase() === 'admin')
      .sort((a, b) => a.firstName.localeCompare(b.firstName));

    return [...owners, ...admins];
  }, [users]);

  return (
    <Stack direction="row" gap={17.5} alignItems="flex-start" py={4}>
      <Box maxWidth={540}>
        <Typography variant="h3" fontWeight={700} mb={1}>
          You need permission to view and edit Roles & Permissions
        </Typography>
        <Typography variant="body2" mb={3} color="text.secondary">
          Contact the instance owner or administrators listed below to upgrade
          your role to Admin or Owner for this capability.
        </Typography>
        <List>
          {ownersAndAdmins?.map((user) => (
            <ListItem
              key={user.ZUID}
              dense
              disableGutters
              sx={{
                borderBottom: '1px solid',
                borderColor: 'border',
              }}
            >
              <ListItemAvatar>
                <Avatar
                  alt={`${user.firstName} ${user.lastName}`}
                  src={`https://www.gravatar.com/avatar/${hashMD5(
                    user.email || '',
                  )}?s=40`}
                />
              </ListItemAvatar>
              <ListItemText
                primary={`${user.firstName} ${user.lastName}`}
                primaryTypographyProps={{
                  sx: {
                    color: 'text.primary',
                  },
                }}
                secondary={`${user.role.name} â€¢ ${user.email}`}
              />
            </ListItem>
          ))}
        </List>
      </Box>
      <Box
        component="img"
        src={shield?.src}
        alt={shield?.title}
        loading="lazy"
      />
    </Stack>
  );
};
