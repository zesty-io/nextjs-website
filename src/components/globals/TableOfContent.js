import { Link, Stack, Typography, useTheme } from '@mui/material';
import React, { memo, useEffect, useState } from 'react';

const Main = ({ data, title = '' }) => {
  const overview = {
    label: title,
    name: title,
    href: `#${title}`,
    originalName: title,
  };
  const newData = [overview, ...data];
  const [currentHash, setcurrentHash] = useState(`#${title}`);
  const handleHashChange = () => {
    setcurrentHash(window.location.hash);
  };
  const theme = useTheme();

  useEffect(() => {
    window.addEventListener('hashchange', handleHashChange);

    return () => {
      window.removeEventListener('hashchange', handleHashChange);
    };
  }, []);

  return (
    <Stack height={1} width={1} data-testid="table-of-contents">
      <Typography variant="button" color={'black'} fontWeight={700} pl={'20px'}>
        On this Page
      </Typography>
      <Stack px={0}>
        {newData.map((e) => {
          const active = currentHash === e.href ? true : false;
          return (
            <Link
              href={e.href}
              sx={{
                borderLeft: `3px solid ${
                  active ? theme?.palette?.zesty?.zestyOrange : '#ffd5c1'
                }`,
                fontWeight: active ? '600' : '400',
                fontSize: '8px',
                textDecoration: active ? 'none !important' : 'none',
                backgroundColor: active ? '#ffd6c4' : 'white',
                color: active
                  ? `${theme?.palette?.zesty?.zestyOrange} !important`
                  : '#6b7280',
                borderRadius: '0 5px 5px 0',
                pl: '20px',
                pr: '5px',
                py: '5px',
                '&:hover': {
                  color: '#333333',
                  textDecoration: 'underline #333333',
                  borderLeft: `3px solid ${theme?.palette?.zesty?.zestyOrange}`,
                },
              }}
            >
              <Typography
                variant="button"
                whiteSpace={'normal'}
                data-testid="link-name"
              >
                {e.originalName}
              </Typography>
            </Link>
          );
        })}
      </Stack>
    </Stack>
  );
};

export const TableOfContent = memo(Main);
