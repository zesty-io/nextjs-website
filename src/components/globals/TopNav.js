import React from 'react';
import PropTypes from 'prop-types';
import Box from '@mui/material/Box';
import Link from '@mui/material/Link';
import { zestyLink } from 'lib/zestyLink';
import ThemeModeToggler from './ThemeModeToggler';

const TopNav = ({ hideNav, nav, colorInvert = false }) => {
  return (
    <Box
      display={hideNav ? 'none' : 'flex'}
      justifyContent={'flex-end'}
      alignItems={'center'}
    >
      <Box marginRight={{ xs: 1, sm: 2 }}>
        <Link
          underline="none"
          component="a"
          href="/docs"
          color={colorInvert ? 'common.white' : 'text.primary'}
        >
          Docs
        </Link>
      </Box>
      <Box marginRight={{ xs: 1, sm: 2 }}>
        <Link
          underline="none"
          component="a"
          href="https://www.github.com/zesty-io"
          color={colorInvert ? 'common.white' : 'text.primary'}
          sx={{ display: 'flex', alignItems: 'center' }}
        >
          Github
        </Link>
      </Box>

      <Box marginRight={{ xs: 1, sm: 2 }}>
        <Link
          underline="none"
          component="a"
          href={zestyLink(nav, '7-cec987fcf5-9bht2z')}
          color={colorInvert ? 'common.white' : 'text.primary'}
          sx={{ display: 'flex', alignItems: 'center' }}
        >
          Contact Us
        </Link>
      </Box>
      <Box>
        <ThemeModeToggler />
      </Box>
    </Box>
  );
};

TopNav.propTypes = {
  colorInvert: PropTypes.bool,
};

export default TopNav;
