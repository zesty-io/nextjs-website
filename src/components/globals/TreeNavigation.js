import React, { memo } from 'react';
import { TreeItem, TreeView } from '@mui/lab';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import { useRouter } from 'next/router';
import NextLink from 'next/link';
import { Typography, useTheme } from '@mui/material';

const slashedDocsChecker = (href, isDocs = false) => {
  if (isDocs) {
    if (href.includes('/docs')) {
      return href;
    } else {
      return `/docs/${href}`;
    }
  }

  return href;
};

const GetTree = ({ data = [] }) => {
  const router = useRouter();
  const isDocs = router?.asPath?.includes('/docs/');

  return data.map((e) => {
    if (e?.children?.length !== 0 && Array.isArray(e?.children)) {
      return (
        <TreeItem
          nodeId={e.uri}
          sx={{
            my: 0.5,
            color: '#6B7280',
          }}
          label={
            <NextLink
              href={slashedDocsChecker(e.uri, isDocs)}
              variant="p"
              color={'inherit'}
              sx={{
                textDecoration: 'none',
              }}
            >
              <Typography
                variant="body1"
                py={0.5}
                title={e.title}
                className="algolia-lvl0"
                component={'h3'}
              >
                {e.title || e.name}
              </Typography>
            </NextLink>
          }
        >
          <GetTree data={e.children} />
        </TreeItem>
      );
    } else {
      return (
        <TreeItem
          nodeId={e.uri}
          sx={{
            my: 0.5,
            color: '#6B7280',
          }}
          label={
            <NextLink
              href={slashedDocsChecker(e.uri, isDocs)}
              variant="p"
              color={'inherit'}
              sx={{
                textDecoration: 'none',
                my: 0.1,
              }}
            >
              <Typography variant="body1" py={0.5} title={e.title}>
                {e.title || e.name}
              </Typography>
            </NextLink>
          }
        ></TreeItem>
      );
    }
  });
};

const Main = ({ data = [] }) => {
  const router = useRouter();
  const url = `${router?.asPath}`;
  const parts = url.split('/');
  parts.splice(3, 1);
  const expanded = [parts.join('/')];
  const theme = useTheme();
  return (
    <TreeView
      data-testid="navigation-tree"
      defaultCollapseIcon={<ExpandMoreIcon />}
      defaultExpandIcon={<ChevronRightIcon />}
      sx={{
        width: 1,
        overflowY: 'auto',

        '& .MuiTreeItem-content.Mui-focused, & .MuiTreeItem-content.Mui-selected, & .MuiTreeItem-content.Mui-selected.Mui-focused':
          {
            bgcolor: '#FFD6C4',
            color: theme?.palette?.zesty?.zestyOrange,
            fontWeight: 'bold !important',
            borderRadius: '5px',
          },
        '& .MuiTreeItem-content:hover': {
          bgcolor: '#F2F2F2',
          color: '#333333',
          fontWeight: '800 !important',
          borderRadius: '5px',
        },
      }}
      multiSelect
      selected={url}
      defaultSelected={url}
      defaultExpanded={expanded}
    >
      <GetTree data={data} />
    </TreeView>
  );
};
export const TreeNavigation = memo(Main);
