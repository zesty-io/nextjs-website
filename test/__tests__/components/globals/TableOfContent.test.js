// main.test.js
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { TableOfContent } from '../../../../src/components/globals/TableOfContent';

const data = [
  {
    value: 'understanding-content-zestyio',
    label: 'Understanding Content Zesty.io',
    name: 'Understanding Content Zesty.io',
    url: '#understanding-content-zestyio',
    href: '#understanding-content-zestyio',
  },
  {
    value: 'content-editing',
    label: 'Content Editing',
    name: 'Content Editing',
    url: '#content-editing',
    href: '#content-editing',
  },
  {
    value: 'visual-layout',
    label: 'Visual Layout',
    name: 'Visual Layout',
    url: '#visual-layout',
    href: '#visual-layout',
  },
  {
    value: 'versioning',
    label: 'Versioning',
    name: 'Versioning',
    url: '#versioning',
    href: '#versioning',
  },
  {
    value: 'environment',
    label: 'Environment',
    name: 'Environment',
    url: '#environment',
    href: '#environment',
  },
  {
    value: 'publishing',
    label: 'Publishing',
    name: 'Publishing',
    url: '#publishing',
    href: '#publishing',
  },
  {
    value: 'workflow-requests',
    label: 'Workflow Requests',
    name: 'Workflow Requests',
    url: '#workflow-requests',
    href: '#workflow-requests',
  },
  {
    value: 'multilanguage-support',
    label: 'Multi-Language Support',
    name: 'Multi-Language Support',
    url: '#multilanguage-support',
    href: '#multi-language-support',
  },
  {
    value: 'api-access',
    label: 'API Access',
    name: 'API Access',
    url: '#api-access',
    href: '#api-access',
  },
];

describe('TableOfContent test', () => {
  test('toc test', () => {
    const { getByTestId } = render(<TableOfContent data={data} />);
    const item = getByTestId('table-of-contents');
    expect(item).toBeInTheDocument();

    expect(item).toMatchSnapshot();
  });
});
