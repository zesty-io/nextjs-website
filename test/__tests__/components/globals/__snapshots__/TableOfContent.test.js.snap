// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableOfContent test toc test 1`] = `
<div
  class="MuiStack-root css-1a8uuxu-MuiStack-root"
  data-testid="table-of-contents"
>
  <span
    class="MuiTypography-root MuiTypography-button css-1armj8x-MuiTypography-root"
  >
    On this Page
  </span>
  <div
    class="MuiStack-root css-p8ma2d-MuiStack-root"
  >
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1wfumw2-MuiTypography-root-MuiLink-root"
      href="#"
    >
      <span
        class="MuiTypography-root MuiTypography-button css-r5ijk3-MuiTypography-root"
        data-testid="link-name"
      />
    </a>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ja28j0-MuiTypography-root-MuiLink-root"
      href="#understanding-content-zestyio"
    >
      <span
        class="MuiTypography-root MuiTypography-button css-r5ijk3-MuiTypography-root"
        data-testid="link-name"
      />
    </a>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ja28j0-MuiTypography-root-MuiLink-root"
      href="#content-editing"
    >
      <span
        class="MuiTypography-root MuiTypography-button css-r5ijk3-MuiTypography-root"
        data-testid="link-name"
      />
    </a>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ja28j0-MuiTypography-root-MuiLink-root"
      href="#visual-layout"
    >
      <span
        class="MuiTypography-root MuiTypography-button css-r5ijk3-MuiTypography-root"
        data-testid="link-name"
      />
    </a>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ja28j0-MuiTypography-root-MuiLink-root"
      href="#versioning"
    >
      <span
        class="MuiTypography-root MuiTypography-button css-r5ijk3-MuiTypography-root"
        data-testid="link-name"
      />
    </a>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ja28j0-MuiTypography-root-MuiLink-root"
      href="#environment"
    >
      <span
        class="MuiTypography-root MuiTypography-button css-r5ijk3-MuiTypography-root"
        data-testid="link-name"
      />
    </a>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ja28j0-MuiTypography-root-MuiLink-root"
      href="#publishing"
    >
      <span
        class="MuiTypography-root MuiTypography-button css-r5ijk3-MuiTypography-root"
        data-testid="link-name"
      />
    </a>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ja28j0-MuiTypography-root-MuiLink-root"
      href="#workflow-requests"
    >
      <span
        class="MuiTypography-root MuiTypography-button css-r5ijk3-MuiTypography-root"
        data-testid="link-name"
      />
    </a>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ja28j0-MuiTypography-root-MuiLink-root"
      href="#multi-language-support"
    >
      <span
        class="MuiTypography-root MuiTypography-button css-r5ijk3-MuiTypography-root"
        data-testid="link-name"
      />
    </a>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ja28j0-MuiTypography-root-MuiLink-root"
      href="#api-access"
    >
      <span
        class="MuiTypography-root MuiTypography-button css-r5ijk3-MuiTypography-root"
        data-testid="link-name"
      />
    </a>
  </div>
</div>
`;
